File: app/src/main/java/com/google/android/leanbacklauncher/notifications/PartnerAdapter.java
package com.google.android.leanbacklauncher.notifications;

import android.app.PendingIntent;
import android.content.Context;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.google.android.leanbacklauncher.R;
import com.google.android.leanbacklauncher.apps.AppsAdapter.AppBannerViewHolder;
import com.google.android.leanbacklauncher.core.LaunchException;
import com.google.android.leanbacklauncher.util.Util;
import com.google.android.tvrecommendations.TvRecommendation;

public class PartnerAdapter extends NotificationsServiceAdapter<PartnerBannerViewHolder> {
    private final BlacklistListener mListener;

    static final class PartnerBannerViewHolder extends AppBannerViewHolder {
        private PendingIntent mIntent;

        public PartnerBannerViewHolder(View v) {
            super(v, null);
        }

        public void init(CharSequence title, Drawable banner, PendingIntent intent, int launchColor) {
            super.init(title, banner, launchColor);
            this.mIntent = intent;
        }

        protected void performLaunch() {
            if (this.mIntent != null) {
                try {
                    Util.startActivity(this.mCtx, this.mIntent);
                } catch (Throwable t) {
                    LaunchException launchException = new LaunchException("Could not launch partner intent", t);
                }
            } else {
                throw new LaunchException("No partner intent to launch: " + getPackageName());
            }
        }
    }

    public PartnerAdapter(Context context, BlacklistListener listener) {
        super(context, 15000, 60000);
        this.mListener = listener;
    }

    public int getItemViewType(int position) {
        return 2;
    }

    public PartnerBannerViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new PartnerBannerViewHolder(LayoutInflater.from(this.mContext).inflate(R.layout.app_banner, parent, false));
    }

    public void onBindViewHolder(PartnerBannerViewHolder appHolder, int position) {
        if (position < getItemCount()) {
            TvRecommendation recommendation = getRecommendation(position);
            appHolder.init(recommendation.getTitle(), new BitmapDrawable(this.mContext.getResources(), recommendation.getContentImage()), recommendation.getContentIntent(), recommendation.getColor());
        }
    }

    protected boolean isPartnerClient() {
        return true;
    }

    protected void onNewRecommendation(TvRecommendation rec) {
        String pkgName = rec.getReplacedPackageName();
        if (!TextUtils.isEmpty(pkgName) && this.mListener != null) {
            this.mListener.onPackageBlacklisted(pkgName);
        }
    }

    protected void onRecommendationRemoved(TvRecommendation rec) {
        String pkgName = rec.getReplacedPackageName();
        if (!TextUtils.isEmpty(pkgName) && this.mListener != null) {
            this.mListener.onPackageUnblacklisted(pkgName);
        }
    }
}

File: app/src/main/java/com/amazon/tv/leanbacklauncher/notifications/PartnerAdapter.kt
package com.amazon.tv.leanbacklauncher.notifications

import android.app.PendingIntent
import android.content.Context
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.amazon.tv.leanbacklauncher.BuildConfig
import com.amazon.tv.leanbacklauncher.R
import com.amazon.tv.leanbacklauncher.apps.AppsAdapter.AppBannerViewHolder
import com.amazon.tv.leanbacklauncher.core.LaunchException
import com.amazon.tv.leanbacklauncher.notifications.PartnerAdapter.PartnerBannerViewHolder
import com.amazon.tv.leanbacklauncher.util.Util.startActivity
import com.amazon.tv.tvrecommendations.TvRecommendation

class PartnerAdapter(context: Context, private val mListener: BlacklistListener?) :
    NotificationsServiceAdapter<PartnerBannerViewHolder?>(context, 15000, 60000) {
    class PartnerBannerViewHolder(v: View?) : AppBannerViewHolder(v!!, null) {
        private var mIntent: PendingIntent? = null
        private val TAG by lazy { if (BuildConfig.DEBUG) ("[*]" + javaClass.simpleName).take(21) else javaClass.simpleName }

        fun init(
            title: CharSequence?,
            banner: Drawable?,
            intent: PendingIntent?,
            launchColor: Int
        ) {
            super.init(title, banner, launchColor)
            mIntent = intent
        }

        override fun performLaunch() {
            if (mIntent != null) {
                try {
                    startActivity(mContext, mIntent!!)
                } catch (e: Exception) {
                    e.message?.let { Log.e(TAG, "Could not launch partner intent, $it ") }
                }
            } else {
                throw LaunchException("No partner intent to launch: $packageName")
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        return 2
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PartnerBannerViewHolder {
        return PartnerBannerViewHolder(
            LayoutInflater.from(mContext).inflate(R.layout.app_banner, parent, false)
        )
    }

    override fun onBindViewHolder(appHolder: PartnerBannerViewHolder, position: Int) {
        if (position < itemCount) {
            val recommendation = getRecommendation(position)
            recommendation?.let {
                appHolder.init(
                    recommendation.title,
                    BitmapDrawable(mContext.resources, recommendation.contentImage),
                    recommendation.contentIntent,
                    recommendation.color
                )
            }
        }
    }

    override val isPartnerClient: Boolean
        get() = true

    override fun onNewRecommendation(rec: TvRecommendation) {
        val pkgName = rec.replacedPackageName
        if (pkgName.isNotEmpty() && mListener != null) {
            mListener.onPackageBlacklisted(pkgName)
        }
    }

    override fun onRecommendationRemoved(rec: TvRecommendation?) {
        val pkgName = rec?.replacedPackageName
        if (!pkgName.isNullOrEmpty() && mListener != null) {
            mListener.onPackageUnblacklisted(pkgName)
        }
    }
}